### ISSUE TITLE
Consider @var annotation for type inference when deserializing 

### ISSUE THREAD

**Issue by goetas** [schmittjoh/serializer#893]
| Q                | A
| ---------------- | -----
| Bug report?      | no
| Feature request? | yes
| BC Break report? | no
| RFC?             | yes

**Comment by goetas** [comment]
Not really fun of it, I prefer explicit types. 

Closing. If someone has valid points, i'm ready to re-consider it.

**Comment by Majkl578** [comment]
:-1:

**Comment by bdsl** [comment]
I was about to make the same feature request as this. `@var` seems just as explicit as @Serializer\Type("string"), and `@var` has the advantage that many different tools will read it. Most of the time we want the types set by the deserialise function to match types understood by IDEs, static analysis tools and documentation generating tools.

To make this more useful it would be good to understand `use` statements for namespace resolution, and if #1090 is implemented nullable types.

**Comment by Majkl578** [comment]
Explicit `@Type` will be even more annoying with PHP 7.4's typed properties.

**Comment by goetas** [comment]
When typed properties will land in PHP, the `@var` annotation will be obsolete, then inferring the type from the property reflection will make perfectly sense.
In that case will be happy to accept a PR with that feature.

**Comment by bdsl** [comment]
Even if 7.4 `@var` won't be entirely obsolete, because people use it with types that are not understood by the PHP engine - e.g. `@var array<\DateTimeInterface>`, `@var int|string`, `@var mixed` etc. The first of these is also supported by the existing [@Type annotation](https://jmsyst.com/libs/serializer/master/reference/annotations#type).

But I might be interested in making that PR for 7.4 typed properties support some time a bit nearer to 7.4 release date.

**Comment by Majkl578** [comment]
@goetas Not really sure what you mean by your comment - `@var` is already substituing absense of the typed properties and therefore supporting it makes sense already.

**Comment by Majkl578** [comment]
Also agreed with @bdsl - 7.4 won't make `@var` obsolete by far - no union types, no generics.

**Comment by goetas** [comment]
Alright, I see that this feature has people that could use it. There are still many questions to be answered on how to implement it, some are: 

- should this be considered only when using the annotation driver?
- if no, what happens for those that have the annotation driver disabled
- should the argumentation used to close https://github.com/schmittjoh/JMSSerializerBundle/pull/427 be valid (at least partailly even in this case?)

**Comment by Majkl578** [comment]
> should this be considered only when using the annotation driver?

No, Doctine Annotations are not a parser for phpDoc, they're not compatible.

> if no, what happens for those that have the annotation driver disabled

N/A, this feature is independent on the annotation driver.

**Comment by goetas** [comment]
implemented as separate driver,  sounds a good idea to me!

**Comment by magnetik** [comment]
I'm not sure : is inferring types from PHP 7.4 types properties in the scope of this issue? Or is this covered elsewhere?

**Comment by Tobion** [comment]
This has been implemented for both var annotations as well as property types in #1214 and #1192. So this can be closed.
